### PR TITLE
Homework 3, Attempt 1 correct branching

### PR DESCRIPTION
**Pull request by EvgeniiKlimenko**

### CODE REVIEW

File: src/test/java/hw3/pages/DifferentElementsPage.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package hw3.pages;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+import java.util.List;
+
+public class DifferentElementsPage {
```
Comment (HronoSF) [inline src/test/java/hw3/pages/DifferentElementsPage.java:10]: Create AbstractPage.
Make this page extends AbstractPage.

Move to AbstractPage page object initialization and WebDriver field

File: src/test/java/hw3/pages/IndexPage.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class IndexPage {
```
Comment (HronoSF) [inline src/test/java/hw3/pages/IndexPage.java:12]: Create AbstractPage.
Make this page extends AbstractPage.

Move to AbstractPage page object initialization and WebDriver field

File: src/test/java/hw3/pages/IndexPage.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * Use only when content is switched to iFrame with switchToIFrameWButton()
+     * @return button within iFrame
+     */
+    public WebElement getButtonFromIFrame() {
+        return myDriver.findElement(By.xpath("/html/body/span/div/div/input"));
+    }
+
+    public void goToDifferentElementsPage() {
+        WebElement headerListTop = myDriver.findElement(By.xpath("//*[@class='uui-navigation nav navbar-nav m-l8']"));
+        myDriver.findElement(By.xpath("//header/div/nav/ul[1]/li[3]")).click();
```
Comment (HronoSF) [inline src/test/java/hw3/pages/IndexPage.java:79]: fix locators

File: src/test/java/hw3/pages/IndexPage.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    @FindBy(id = "name")
+    private WebElement user;
+
+    @FindBy(id = "password")
+    private WebElement password;
+
+    @FindBy(id = "login-button")
+    private WebElement submitBtn;
+
+    @FindBy(xpath = ("/html/body/header/div/nav/ul[1]/li[position()<5]/a"))
```
Comment (HronoSF) [inline src/test/java/hw3/pages/IndexPage.java:24]: fix locators

File: src/test/java/hw3/pages/IndexPage.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    List<WebElement> headerHorizontalList;
+
+    @FindBy(className = ("benefit-icon"))
+    List<WebElement> benefitImagesList;
+
+    @FindBy(xpath = ("/html/body/div/div[2]/main/div[2]/div[2]/div[position()<5]/div/span"))
```
Comment (HronoSF) [inline src/test/java/hw3/pages/IndexPage.java:30]: fix locators

File: src/test/java/hw3/pages/IndexPage.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    List<WebElement> benefitTextsList;
+
+    @FindBy(xpath = ("/html/body/div/div[1]/div/div[1]/div/div[1]/ul/li[position()<6]/a/span"))
```
Comment (HronoSF) [inline src/test/java/hw3/pages/IndexPage.java:33]: fix locators

File: src/test/java/hw3/pages/DifferentElementsPage.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    private WebDriver myDriver;
+
+    @FindBy(xpath = ("/html/body/div/div[3]/div/div[1]/div/div[1]/section[1]/div[2]/div/ul/li"))
```
Comment (HronoSF) [inline src/test/java/hw3/pages/DifferentElementsPage.java:13]: fix locators

File: src/test/java/hw3/ex2/Exercise2Test.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        //5. Service -> Different Elements Page
+        indexPage.goToDifferentElementsPage();
+        Assert.assertEquals(driver.getTitle(), "Different Elements");
+
+        // How to handle with such elements like in 6-7-8 ???
+        //6. Select checkboxes
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[2]/label[1]")).click();
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[2]/label[3]")).click();
+
+        //7. Select radio button
```
Comment (HronoSF) [inline src/test/java/hw3/ex2/Exercise2Test.java:37]: all WebElement work must be in page object, move to PO

File: src/test/java/hw3/ex2/Exercise2Test.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        //4. User name
+        Assert.assertTrue(indexPage.isLoggedUserCorrect("ROMAN IOVLEV"));
+
+        //5. Service -> Different Elements Page
+        indexPage.goToDifferentElementsPage();
+        Assert.assertEquals(driver.getTitle(), "Different Elements");
+
+        // How to handle with such elements like in 6-7-8 ???
+        //6. Select checkboxes
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[2]/label[1]")).click();
```
Comment (HronoSF) [inline src/test/java/hw3/ex2/Exercise2Test.java:34]: all WebElement work must be in page object, move to PO

File: src/test/java/hw3/ex2/Exercise2Test.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[2]/label[3]")).click();
```
Comment (HronoSF) [inline src/test/java/hw3/ex2/Exercise2Test.java:35]: all WebElement work must be in page object, move to PO

File: src/test/java/hw3/ex2/Exercise2Test.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+        //7. Select radio button
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[3]/label[4]")).click();
+
+        //8. Select dropdown
+        driver.findElement(By.xpath("/html/body/div/div[2]/main/div[2]/div/div[4]/select/option[4]")).click();
```
Comment (HronoSF) [inline src/test/java/hw3/ex2/Exercise2Test.java:41]: all WebElement work must be in page object, move to PO

File: src/test/java/hw3/ex1/Exercise1Test.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.List;
+
+
+public class Exercise1Test extends AbstractBaseTest {
+
+    private ArrayList<String> menuList;
+
+    @BeforeTest
+    public void initialize() {
+        menuList = new ArrayList<>(5);
```
Comment (HronoSF) [inline src/test/java/hw3/ex1/Exercise1Test.java:20]: to enum

File: pom.xml [hunk 13]
```diff
@@ -8,6 +8,69 @@
     <artifactId>AQAhw1</artifactId>
     <version>1.0-SNAPSHOT</version>
 
+    <profiles>
```
Comment (HronoSF) [inline pom.xml:11]: move Page objects to main package

File: src/test/java/hw3/AbstractBaseTest.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    @BeforeClass
+    public void setUp() {
+        WebDriverManager.chromedriver().browserVersion("89.0.4389.82").setup();
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.manage().timeouts().implicitlyWait(Duration.ofMillis(5000));
+        driver.manage().timeouts().pageLoadTimeout(Duration.ofMillis(10000));
+
+        props = new Properties();
+        try {
```
Comment (HronoSF) [inline src/test/java/hw3/AbstractBaseTest.java:29]: you will read properties before each test class. Do you need it?
You should create class in utils folder and use properties from it (via static for example)

File: src/test/java/hw3/ex1/Exercise1Test.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+public class Exercise1Test extends AbstractBaseTest {
+
+    @Test
+    public void exerciseOneTest() {
+        driver.get(URL);
+        IndexPage indexPage = new IndexPage(driver);
+        Assert.assertEquals(driver.getTitle(), "Home Page");
+
+        //3. Perform login
+        indexPage.login(props.getProperty("user.roman.login"), props.getProperty("user.roman.password"));
```
Comment (HronoSF) [inline src/test/java/hw3/ex1/Exercise1Test.java:22]: do not trigger each time props.getProperty("user.roman.login") or etc. Code duplcication is bad. Create field in AbstractBaseTest and use it

File: pom.xml [hunk 16]
```diff
@@ -8,12 +8,28 @@
     <artifactId>AQAhw1</artifactId>
     <version>1.0-SNAPSHOT</version>
 
+    <properties>
```
Comment (HronoSF) [inline pom.xml:11]: where is pages?
image

File: pom.xml [hunk 17]
```diff
@@ -68,6 +84,18 @@
             <version>4.0.0-beta-1</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-api</artifactId>
+            <version>4.0.0-beta-1</version>
```
Comment (komyagin) [inline pom.xml:90]: Version should be moved to properties

File: pom.xml [hunk 18]
```diff
@@ -68,6 +84,18 @@
...
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-support</artifactId>
+            <version>4.0.0-beta-1</version>
```
Comment (komyagin) [inline pom.xml:96]: Version should be moved to properties

File: src/main/java/pages/DifferentElementsPage.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        myDriver.findElement(By.xpath("//label[contains(. ,'Wind')]/input")).click(); // Wind
+    }
+
+    public void selectRadioButton() {
+        myDriver.findElement(By.xpath("//label[contains(. ,'Selen')]/input")).click(); // Selen
+    }
+
+    public void selectFromDropdownMenu() {
+        myDriver.findElement(By.xpath("//*[@class='colors']/select/option[4]")).click(); // Yellow
+    }
```
Comment (komyagin) [inline src/main/java/pages/DifferentElementsPage.java:34]: Three methods with the same function should use parameters. One method - one action (read about SOLID principles)

File: src/main/java/pages/IndexPage.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    @FindBy(id = "name")
+    private WebElement userField;
+
+    @FindBy(id = "password")
+    private WebElement passwordField;
+
+    @FindBy(id = "login-button")
+    private WebElement submitBtn;
+
+    @FindBy(xpath = ("//*[@class='uui-navigation nav navbar-nav m-l8']/li[position()<5]/a"))
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:22]: Are you sure that is the best locator?

File: src/main/java/pages/IndexPage.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    List<WebElement> headerHorizontalList;
+
+    @FindBy(className = ("benefit-icon"))
+    List<WebElement> benefitImagesList;
+
+    @FindBy(xpath = ("//*[@class='col-sm-3']/div/span"))
+    List<WebElement> benefitTextsList;
+
+    @FindBy(xpath = ("//*[@class='sidebar-menu']/li[position()<6]/a/span"))
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:31]: Are you sure that is the best locator?

File: src/main/java/pages/IndexPage.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    List<WebElement> sideBarMenuList;
+
+    public IndexPage(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(driver, this);
+    }
+
+    public void login(String login, String pass) {
+        myDriver.findElements(By.className("dropdown-toggle")).get(1).click();
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:40]: You should move By.className("dropdown-toggle") to a variable. Are you sure that get(1) is the only possibility to get element?

File: src/main/java/pages/IndexPage.java [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+        userField.sendKeys(login);
+        this.passwordField.sendKeys(pass);
+        submitBtn.click();
+    }
+
+    public boolean isLoggedUserCorrect(String userExpectedName) {
+        WebElement userName = myDriver.findElement(By.id("user-name"));
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:47]: You should move By.id("user-name") to a variable

File: src/main/java/pages/IndexPage.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    public List<WebElement> getHeaderHorizList() { return headerHorizontalList; }
+
+    public List<WebElement> getBenefitImagesList() { return benefitImagesList; }
+
+    public List<WebElement> getBenefitTextsList() { return benefitTextsList; }
+
+    public List<WebElement> getSideBarMenuList() { return sideBarMenuList;}
+
+    public WebElement getIFrameByName(String frameName) {
+        return myDriver.findElement(By.id(frameName));
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:60]: You should move By.id(frameName) to a variable

File: src/main/java/pages/IndexPage.java [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    public void switchToDefaultContent() {
+        myDriver.switchTo().defaultContent();
+    }
+
+    /**
+     * Use only when content is switched to iFrame with switchToIFrameWButton()
+     * @return button within iFrame
+     */
+    public WebElement getButtonFromIFrame() {
+        return myDriver.findElement(By.xpath("/html/body/span/div/div/input"));
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:72]: It's a bad practice to use an absolute path, only if you don't have another option. And you should move By.xpath to a variable

File: src/main/java/pages/IndexPage.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    public void goToDifferentElementsPage() {
+        WebElement headerListTop = myDriver.findElement(By.xpath("//*[@class='uui-navigation nav navbar-nav m-l8']"));
+        myDriver.findElement(By.xpath("//header/div/nav/ul[1]/li[3]")).click();
+        headerListTop.findElement(By.xpath("//li[3]/ul/li[8]/a")).click();
```
Comment (komyagin) [inline src/main/java/pages/IndexPage.java:78]: Move locators to variables. And maybe we can move WebElement to the top? Is it necessary to declare WebElement here? Are you that these locators are the best?

File: src/main/java/utils/PropertiesHolder.java [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Properties;
+
+public class PropertiesHolder {
+
+    public static final Properties PROPS;
+
+    static {
+        PROPS = new Properties();
+        try {
+            PROPS.load(new FileInputStream("src/test/resources/userData.properties"));
```
Comment (komyagin) [inline src/main/java/utils/PropertiesHolder.java:14]: Declare a constant for the path

File: src/test/java/hw3/AbstractBaseTest.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+
+    protected WebDriver driver;
+    protected final String URL = "https://jdi-testing.github.io/jdi-light/index.html";
+    protected final String ROMAN_LOGIN = PropertiesHolder.PROPS.getProperty("user.roman.login");
+    protected final String ROMAN_PASSWORD = PropertiesHolder.PROPS.getProperty("user.roman.password");
+    protected final String ROMAN_CORRECT_NAME = PropertiesHolder.PROPS.getProperty("user.roman.correctName");
+
+    @BeforeClass
+    public void setUp() {
+        WebDriverManager.chromedriver().browserVersion("89.0.4389.82").setup();
```
Comment (komyagin) [inline src/test/java/hw3/AbstractBaseTest.java:22]: If you use WebDriverManager why you set the driver version? Is it necessary? If so move the version to constant. But you use PropertiesHolder why don't you use it for version loading?

File: src/test/java/hw3/ex1/Exercise1Test.java [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+        Assert.assertEquals(driver.getTitle(), "Home Page");
+
+        //3. Perform login
+        indexPage.login(ROMAN_LOGIN, ROMAN_PASSWORD);
+
+        //4. User name
+        Assert.assertTrue(indexPage.isLoggedUserCorrect(ROMAN_CORRECT_NAME));
+
+        //5. 4 items on the header
+        Assert.assertEquals(indexPage.getHeaderHorizList().size(), 4);
```
Comment (komyagin) [inline src/test/java/hw3/ex1/Exercise1Test.java:28]: Don't you magic numbers in assertions. It's good practice to declare constant for this.

File: src/test/java/hw3/ex1/Exercise1Test.java [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+        for (WebElement el: indexPage.getHeaderHorizList()) {
+            Assert.assertTrue(MenuList.HOME.checkIsContains(el.getText()));
+        }
+
+        //6. 4 images on the Index Page are displayed
+        Assert.assertEquals(indexPage.getBenefitImagesList().size(), 4);
```
Comment (komyagin) [inline src/test/java/hw3/ex1/Exercise1Test.java:34]: Don't you magic numbers in assertions. It's good practice to declare constant for this.

File: src/test/java/hw3/ex1/Exercise1Test.java [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+        indexPage.goToIFrameWithButton("frame");
+        WebElement frameButton = indexPage.getButtonFromIFrame();
+        Assert.assertTrue(frameButton.isDisplayed());
+
+        //10. Switch back
+        indexPage.switchToDefaultContent();
+
+        //11. Sidebar menu
+        List<WebElement> sidebarList = indexPage.getSideBarMenuList();
+        Assert.assertEquals(sidebarList.size(), 5);
```
Comment (komyagin) [inline src/test/java/hw3/ex1/Exercise1Test.java:62]: Don't you magic numbers in assertions. It's good practice to declare constant for this.

**Review by komyagin — CHANGES_REQUESTED** [review]

**Review by komyagin — APPROVED** [review]
HW#3 is accepted

**Pull request merged**